backend/services/event.py

```diff
@@ -199,3 +220,142 @@ def delete(self, subject: User, id: int) -> None:
+        if existing_registration:
+            return existing_registration
+
+        # Add new object to table and commit changes
+        event_registration_entity = EventRegistrationEntity(
+            user_id=attendee.id, event_id=event.id
+        )
+        self._session.add(event_registration_entity)
+        self._session.commit()
+
+        # Return registration
+        return event_registration_entity.to_model()
+
+    def unregister(self, subject: User, attendee: User, event: EventDetails) -> None:
+        """
+        Delete a user's event registration.
+
+        Args:
+            subject: User performing the unregister action
+            attendee: User whose registration is being deleted
+            event: the event the attendee is unregistering for
+
+        Returns:
+            None in a successful invocation. Idempotent in the case of not registered.
+
+        Raises:
+            UserPermissionException when the user is not authorized to manage the registration.
+        """
+        # Find registration to delete
+        # Permissions for reading/managing registration are enfoced in #get_registration
+        event_registration = self.get_registration(subject, attendee, event)
+
+        # Ensure object exists
+        if event_registration is None:
+            return None
+
+        # Delete object and commit
+        self._session.delete(
+            self._session.get(EventRegistrationEntity, event_registration.id)
+        )
+        self._session.commit()
+
+    def get_registrations_of_user(
+        self, subject: User, user: User, time_range: TimeRange
+    ) -> Sequence[EventRegistration]:
+        """
+        Get a user's registrations to events falling within a given time range.
+
+        Args:
+            subject: The User making the request.
+            user: The User whose registrations are being requested.
+            time_range: The period over which to search for event registrations.
+
+        Returns:
+            Sequence[EventRegistration] event registrations
+
+        Raises:
+            UserPermissionException when the user is requesting the registrations
+            of another user and does not have 'user.event_registrations' permission.
+        """
+        # Feature-specific authorization: User is getting their own registrations
+        # Administrative Permission: user.event_registrations : user/{user_id}
+        if subject.id != user.id:
+            self._permission.enforce(
+                subject,
+                "user.event_registrations",
+                f"user/{user.id}",
+            )
+
+        registration_entities = (
+            self._session.query(EventRegistrationEntity)
+            .where(EventRegistrationEntity.user_id == user.id)
+            .join(EventEntity, EventRegistrationEntity.event_id == EventEntity.id)
+            .where(EventEntity.time >= time_range.start)
+            .where(EventEntity.time < time_range.end)

```

Comment:
I wonder if there is a way to make some sort of helper or query builder in Python that would allow us to more easily compare date ranges! Unsure if this exists though.

Reply:
This is worth looking into, certainly we could write a helper ourself that has a form of something like:

`within_time_range(query, EventEntity.time)`